clients/src/main/java/org/apache/kafka/common/memory/SimpleMemoryPool.java

```diff
@@ -0,0 +1,141 @@
+        boolean success = false;
+        //in strict mode we will only allocate memory if we have at least the size required.
+        //in non-strict mode we will allocate memory if we have _any_ memory available (so available memory
+        //can dip into the negative and max allocated memory would be sizeBytes + maxSingleAllocationSize)
+        long threshold = strict ? sizeBytes : 1;
+        while ((available = availableMemory.get()) >= threshold) {
+            success = availableMemory.compareAndSet(available, available - sizeBytes);
+            if (success)
+                break;
+        }
+
+        if (success) {
+            maybeRecordEndOfDrySpell();
+        } else {
+            if (oomTimeSensor != null) {
+                startOfNoMemPeriod.compareAndSet(0, System.currentTimeMillis());
+            }
+            log.trace("refused to allocate buffer of size {}", sizeBytes);
+            return null;
+        }
+
+        ByteBuffer allocated = ByteBuffer.allocate(sizeBytes);
+        bufferToBeReturned(allocated);
+        return allocated;
+    }
+
+    @Override
+    public void release(ByteBuffer previouslyAllocated) {
+        if (previouslyAllocated == null)
+            throw new IllegalArgumentException("provided null buffer");
+
+        bufferToBeReleased(previouslyAllocated);
+        availableMemory.addAndGet(previouslyAllocated.capacity());
+        maybeRecordEndOfDrySpell();
+    }
+
+    @Override
+    public long size() {
+        return sizeBytes;
+    }
+
+    @Override
+    public long availableMemory() {
+        return availableMemory.get();
+    }
+
+    @Override
+    public boolean isOutOfMemory() {
+        return availableMemory.get() <= 0;
+    }
+
+    @Override
+    public void setOutOfMemoryPeriodSensor(Sensor oomPeriodSensor) {
+        this.oomTimeSensor = oomPeriodSensor;
+    }
+
+    //allows subclasses to do their own bookkeeping (and validation) _before_ memory is returned to client code.
+    protected void bufferToBeReturned(ByteBuffer justAllocated) {
+        log.trace("allocated buffer of size {} ", justAllocated.capacity());
+    }
+
+    //allows subclasses to do their own bookkeeping (and validation) _before_ memory is marked as reclaimed.
+    protected void bufferToBeReleased(ByteBuffer justReleased) {
+        log.trace("released buffer of size {}", justReleased.capacity());
+    }
+
+    @Override
+    public String toString() {
+        long allocated = sizeBytes - availableMemory.get();
+        return "SimpleMemoryPool{" + Utils.formatBytes(allocated) + "/" + Utils.formatBytes(sizeBytes) + " used}";
+    }
+
+    protected void maybeRecordEndOfDrySpell() {
+        if (oomTimeSensor != null) {
+            long startOfDrySpell = startOfNoMemPeriod.getAndSet(0);
+            if (startOfDrySpell != 0) {
+                //how long were we refusing allocation requests for
+                oomTimeSensor.record(System.currentTimeMillis() - startOfDrySpell);
```

Comment:
Could we track this at nano sec level and pass the value as double in ms for better accuracy?